File: scripts/validate-report-links.py
#!/usr/bin/env python3
"""
Report Link Validation Script

Parses a weekly report markdown file, extracts all GitHub PR, GitLab MR,
and Jira links, then verifies each via API calls.

Engineer username mappings are loaded from data/team-config.json (not hardcoded).

Usage:
  python3 scripts/validate-report-links.py                              # latest report
  python3 scripts/validate-report-links.py path/to/report.md            # specific file
  python3 scripts/validate-report-links.py --verbose                    # show all checks
"""

import argparse
import json
import os
import re
import subprocess
import sys
from pathlib import Path
from typing import Dict, List, NamedTuple, Optional
from urllib.parse import urlparse
from urllib.request import Request, urlopen
from urllib.error import HTTPError, URLError

GREEN = '\033[92m'
YELLOW = '\033[93m'
RED = '\033[91m'
BLUE = '\033[94m'
DIM = '\033[2m'
RESET = '\033[0m'


def load_engineer_maps(config_path: Path) -> tuple:
    if not config_path.exists():
        print(f"{YELLOW}WARN: team config not found at {config_path}, author checks disabled{RESET}")
        return {}, {}
    with open(config_path) as f:
        config = json.load(f)
    github_map = {e["name"]: e["github"] for e in config["engineers"]}
    gitlab_map = {e["name"]: e["gitlab"] for e in config["engineers"]}
    return github_map, gitlab_map


class LinkInfo(NamedTuple):
    line_num: int
    link_text: str
    url: str
    engineer: Optional[str]


class LinkResult(NamedTuple):
    link: LinkInfo
    status: str  # "ok", "error", "warning"
    message: str


class ReportLinkValidator:
    def __init__(self, project_root: Path, verbose: bool = False):
        self.project_root = project_root
        self.verbose = verbose
        self.results: List[LinkResult] = []
        self._github_token = None
        self._gitlab_token = None
        self._load_tokens()
        config_path = project_root / "data" / "team-config.json"
        self._engineer_github, self._engineer_gitlab = load_engineer_maps(config_path)

    def _load_tokens(self):
        env_path = self.project_root / ".env"
        if not env_path.exists():
            return
        with open(env_path) as f:
            for line in f:
                line = line.strip()
                if line.startswith("#") or "=" not in line:
                    continue
                key, _, val = line.partition("=")
                if key == "GITHUB_PAT":
                    self._github_token = val
                elif key == "GITLAB_PAT":
                    self._gitlab_token = val

    def _api_get(self, url: str, token: str, token_type: str = "token") -> Optional[dict]:
        headers = {"Accept": "application/json"}
        if token_type == "token":
            headers["Authorization"] = f"token {token}"
        elif token_type == "bearer":
            headers["Authorization"] = f"Bearer {token}"
        try:
            req = Request(url, headers=headers)
            with urlopen(req, timeout=15) as resp:
                return json.loads(resp.read())
        except HTTPError as e:
            if e.code == 404:
                return None
            raise
        except (URLError, TimeoutError):
            return None

    def parse_links(self, file_path: Path) -> List[LinkInfo]:
        with open(file_path) as f:
            lines = f.readlines()

        links: List[LinkInfo] = []
        current_engineer = None

        for line_num, line in enumerate(lines, start=1):
            eng_match = re.match(r'\*\*([A-Z][a-z]+ [A-Z][a-z]+)', line)
            if eng_match:
                current_engineer = eng_match.group(1)

            for m in re.finditer(r'\[([^\]]+)\]\((https?://[^)]+)\)', line):
                link_text, url = m.group(1), m.group(2)
                links.append(LinkInfo(line_num, link_text, url, current_engineer))

        return links

    def classify_links(self, links: List[LinkInfo]) -> Dict[str, List[LinkInfo]]:
        classified = {"github_pr": [], "gitlab_mr": [], "jira": [], "other": []}
        for link in links:
            if "github.com" in link.url and "/pull/" in link.url:
                classified["github_pr"].append(link)
            elif "gitlab" in link.url and "/merge_requests/" in link.url:
                classified["gitlab_mr"].append(link)
            elif "atlassian.net/browse/" in link.url:
                classified["jira"].append(link)
            else:
                classified["other"].append(link)
        return classified

    def verify_github_pr(self, link: LinkInfo) -> LinkResult:
        m = re.search(r'github\.com/([^/]+)/([^/]+)/pull/(\d+)', link.url)
        if not m:
            return LinkResult(link, "error", "Could not parse GitHub PR URL")

        owner, repo, pr_num = m.group(1), m.group(2), m.group(3)

        text_num = re.search(r'#(\d+)', link.link_text)
        if text_num and text_num.group(1) != pr_num:
            return LinkResult(link, "error",
                f"PR number mismatch: text says #{text_num.group(1)} but URL has /pull/{pr_num}")

        if not self._github_token:
            return LinkResult(link, "warning", "No GITHUB_PAT — skipped API check")

        api_url = f"https://api.github.com/repos/{owner}/{repo}/pulls/{pr_num}"
        data = self._api_get(api_url, self._github_token)

        if data is None:
            return LinkResult(link, "error", f"404 NOT FOUND: {owner}/{repo}/pull/{pr_num}")

        actual_author = data.get("user", {}).get("login", "unknown")
        if link.engineer:
            expected = self._engineer_github.get(link.engineer)
            if expected and actual_author != expected:
                return LinkResult(link, "error",
                    f"WRONG AUTHOR: expected {expected}, got {actual_author}")

        actual_title = data.get("title", "")
        desc_in_text = re.sub(r'^(PR #\d+ - |MR \d+ - )', '', link.link_text)
        desc_in_text = re.sub(r'\s*\([^)]*\)\s*$', '', desc_in_text).strip()
        if desc_in_text and actual_title and desc_in_text.lower() not in actual_title.lower() and actual_title.lower() not in desc_in_text.lower():
            return LinkResult(link, "warning",
                f"Title drift: text='{desc_in_text}' vs actual='{actual_title[:60]}'")

        state = data.get("state", "unknown")
        merged = data.get("merged", False)
        status_str = "merged" if merged else state
        return LinkResult(link, "ok",
            f"{owner}/{repo} #{pr_num} ({actual_author}, {status_str})")

    def verify_gitlab_mr(self, link: LinkInfo) -> LinkResult:
        m = re.search(r'gitlab[^/]*/(.+?)/-/merge_requests/(\d+)', link.url)
        if not m:
            return LinkResult(link, "error", "Could not parse GitLab MR URL")

        project_path, mr_iid = m.group(1), m.group(2)

        text_num = re.search(r'(?:MR |!)(\d+)', link.link_text)
        if text_num and text_num.group(1) != mr_iid:
            return LinkResult(link, "error",
                f"MR number mismatch: text says !{text_num.group(1)} but URL has /merge_requests/{mr_iid}")

        if not self._gitlab_token:
            return LinkResult(link, "warning", "No GITLAB_PAT — skipped API check")

        encoded_path = project_path.replace("/", "%2F")
        api_url = f"https://gitlab.cee.redhat.com/api/v4/projects/{encoded_path}/merge_requests/{mr_iid}"
        data = self._api_get(api_url, self._gitlab_token, token_type="bearer")

        if data is None:
            return LinkResult(link, "error", f"404 NOT FOUND: {project_path}!{mr_iid}")

        actual_author = data.get("author", {}).get("username", "unknown")
        if link.engineer:
            expected = self._engineer_gitlab.get(link.engineer)
            if expected and actual_author != expected:
                return LinkResult(link, "error",
                    f"WRONG AUTHOR: expected {expected}, got {actual_author}")

        actual_title = data.get("title", "")
        state = data.get("state", "unknown")
        return LinkResult(link, "ok",
            f"{project_path} !{mr_iid} ({actual_author}, {state})")

    def verify_jira(self, link: LinkInfo) -> LinkResult:
        m = re.search(r'/browse/([A-Z]+-\d+)', link.url)
        if not m:
            return LinkResult(link, "error", "Could not parse Jira ticket URL")
        ticket = m.group(1)

        text_ticket = re.search(r'([A-Z]+-\d+)', link.link_text)
        if text_ticket and text_ticket.group(1) != ticket:
            return LinkResult(link, "error",
                f"Ticket mismatch: text says {text_ticket.group(1)} but URL has {ticket}")

        return LinkResult(link, "ok", f"{ticket}")

    def validate(self, file_path: Path) -> bool:
        print(f"\n{BLUE}{'='*60}{RESET}")
        print(f"{BLUE}Report Link Validation{RESET}")
        print(f"{BLUE}{'='*60}{RESET}")
        print(f"\nValidating: {file_path}\n")

        links = self.parse_links(file_path)
        classified = self.classify_links(links)

        if classified["github_pr"]:
            print(f"\n{BLUE}GitHub PRs ({len(classified['github_pr'])}){RESET}")
            for link in classified["github_pr"]:
                result = self.verify_github_pr(link)
                self.results.append(result)
                self._print_result(result)

        if classified["gitlab_mr"]:
            print(f"\n{BLUE}GitLab MRs ({len(classified['gitlab_mr'])}){RESET}")
            for link in classified["gitlab_mr"]:
                result = self.verify_gitlab_mr(link)
                self.results.append(result)
                self._print_result(result)

        if classified["jira"]:
            print(f"\n{BLUE}Jira Tickets ({len(classified['jira'])}){RESET}")
            for link in classified["jira"]:
                result = self.verify_jira(link)
                self.results.append(result)
                self._print_result(result)

        return self._print_summary()

    def _print_result(self, result: LinkResult):
        if result.status == "ok":
            if self.verbose:
                print(f"  {GREEN}✓{RESET} Line {result.link.line_num}: {result.message}")
        elif result.status == "warning":
            print(f"  {YELLOW}⚠{RESET} Line {result.link.line_num}: {result.message}")
        elif result.status == "error":
            print(f"  {RED}✗{RESET} Line {result.link.line_num}: {result.link.link_text}")
            print(f"    {RED}{result.message}{RESET}")

    def _print_summary(self) -> bool:
        total = len(self.results)
        ok = sum(1 for r in self.results if r.status == "ok")
        warnings = sum(1 for r in self.results if r.status == "warning")
        errors = sum(1 for r in self.results if r.status == "error")

        print(f"\n{BLUE}{'='*60}{RESET}")
        print(f"Summary: {total} checked, {GREEN}{ok} passed{RESET}, ", end="")
        if warnings:
            print(f"{YELLOW}{warnings} warnings{RESET}, ", end="")
        if errors:
            print(f"{RED}{errors} broken{RESET}")
        else:
            print(f"{GREEN}0 broken{RESET}")
        print(f"{BLUE}{'='*60}{RESET}\n")

        return errors == 0


def find_latest_report(project_root: Path) -> Optional[Path]:
    report_dir = project_root / "data" / "team-wide"
    if not report_dir.exists():
        return None
    reports = sorted(report_dir.glob("weekly-update-*.md"), reverse=True)
    return reports[0] if reports else None


def main():
    parser = argparse.ArgumentParser(
        description="Validate links in weekly team reports"
    )
    parser.add_argument(
        "file",
        nargs="?",
        help="Path to report markdown file (default: latest in data/team-wide/)"
    )
    parser.add_argument(
        "--verbose", "-v",
        action="store_true",
        help="Show all checks including passing ones"
    )
    args = parser.parse_args()

    project_root = Path(__file__).resolve().parent.parent

    if args.file:
        file_path = Path(args.file)
        if not file_path.is_absolute():
            file_path = Path.cwd() / file_path
    else:
        file_path = find_latest_report(project_root)
        if not file_path:
            print(f"{RED}No weekly reports found in data/team-wide/{RESET}")
            sys.exit(1)

    if not file_path.exists():
        print(f"{RED}File not found: {file_path}{RESET}")
        sys.exit(1)

    validator = ReportLinkValidator(project_root, verbose=args.verbose)
    success = validator.validate(file_path)
    sys.exit(0 if success else 1)


if __name__ == "__main__":
    main()

File: agents/weekly-team-update/scripts/validate-report-links.ts
#!/usr/bin/env tsx
import { readFileSync, existsSync } from "node:fs";
import { resolve, dirname } from "node:path";
import { readdirSync } from "node:fs";

const GREEN = "\x1b[92m";
const YELLOW = "\x1b[93m";
const RED = "\x1b[91m";
const BLUE = "\x1b[94m";
const RESET = "\x1b[0m";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface LinkInfo {
  line_num: number;
  link_text: string;
  url: string;
  engineer: string | null;
}

export interface LinkResult {
  link: LinkInfo;
  status: "ok" | "error" | "warning";
  message: string;
}

interface EngineerMaps {
  github: Map<string, string>;
  gitlab: Map<string, string>;
}

// ---------------------------------------------------------------------------
// Config loading
// ---------------------------------------------------------------------------

export function loadEngineerMaps(configPath: string): EngineerMaps {
  if (!existsSync(configPath)) {
    console.log(
      `${YELLOW}WARN: team config not found at ${configPath}, author checks disabled${RESET}`,
    );
    return { github: new Map(), gitlab: new Map() };
  }
  const config = JSON.parse(readFileSync(configPath, "utf-8")) as {
    engineers: { name: string; github: string; gitlab: string }[];
  };
  const github = new Map<string, string>(
    config.engineers.map((e) => [e.name, e.github]),
  );
  const gitlab = new Map<string, string>(
    config.engineers.map((e) => [e.name, e.gitlab]),
  );
  return { github, gitlab };
}

// ---------------------------------------------------------------------------
// Link parsing
// ---------------------------------------------------------------------------

export function parseLinks(content: string): LinkInfo[] {
  const lines = content.split("\n");
  const links: LinkInfo[] = [];
  let currentEngineer: string | null = null;

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];
    const lineNum = i + 1;

    const engMatch = line.match(/^\*\*([A-Z][a-z]+ [A-Z][a-z]+)/);
    if (engMatch) {
      currentEngineer = engMatch[1];
    }

    const linkRe = /\[([^\]]+)\]\((https?:\/\/[^)]+)\)/g;
    let m: RegExpExecArray | null;
    while ((m = linkRe.exec(line)) !== null) {
      links.push({
        line_num: lineNum,
        link_text: m[1],
        url: m[2],
        engineer: currentEngineer,
      });
    }
  }

  return links;
}

export function classifyLinks(links: LinkInfo[]): {
  github_pr: LinkInfo[];
  gitlab_mr: LinkInfo[];
  jira: LinkInfo[];
  other: LinkInfo[];
} {
  const classified = {
    github_pr: [] as LinkInfo[],
    gitlab_mr: [] as LinkInfo[],
    jira: [] as LinkInfo[],
    other: [] as LinkInfo[],
  };
  for (const link of links) {
    if (link.url.includes("github.com") && link.url.includes("/pull/")) {
      classified.github_pr.push(link);
    } else if (
      link.url.includes("gitlab") &&
      link.url.includes("/merge_requests/")
    ) {
      classified.gitlab_mr.push(link);
    } else if (link.url.includes("atlassian.net/browse/")) {
      classified.jira.push(link);
    } else {
      classified.other.push(link);
    }
  }
  return classified;
}

// ---------------------------------------------------------------------------
// Verification
// ---------------------------------------------------------------------------

async function apiGet(
  url: string,
  token: string,
  tokenType: "token" | "bearer" = "token",
): Promise<Record<string, unknown> | null> {
  const headers: Record<string, string> = { Accept: "application/json" };
  if (tokenType === "token") {
    headers.Authorization = `token ${token}`;
  } else {
    headers.Authorization = `Bearer ${token}`;
  }
  try {
    const resp = await fetch(url, {
      headers,
      signal: AbortSignal.timeout(15000),
    });
    if (resp.status === 404) return null;
    if (!resp.ok) return null;
    return (await resp.json()) as Record<string, unknown>;
  } catch {
    return null;
  }
}

export function verifyGithubPrSync(link: LinkInfo): LinkResult {
  const m = link.url.match(/github\.com\/([^/]+)\/([^/]+)\/pull\/(\d+)/);
  if (!m) {
    return { link, status: "error", message: "Could not parse GitHub PR URL" };
  }

  const [, , , prNum] = m;

  const textNum = link.link_text.match(/#(\d+)/);
  if (textNum && textNum[1] !== prNum) {
    return {
      link,
      status: "error",
      message: `PR number mismatch: text says #${textNum[1]} but URL has /pull/${prNum}`,
    };
  }

  return { link, status: "ok", message: `PR #${prNum} — URL format valid` };
}

export function verifyGitlabMrSync(link: LinkInfo): LinkResult {
  const m = link.url.match(/gitlab[^/]*\/(.+?)\/-\/merge_requests\/(\d+)/);
  if (!m) {
    return { link, status: "error", message: "Could not parse GitLab MR URL" };
  }

  const [, , mrIid] = m;

  const textNum = link.link_text.match(/(?:MR |!)(\d+)/);
  if (textNum && textNum[1] !== mrIid) {
    return {
      link,
      status: "error",
      message: `MR number mismatch: text says !${textNum[1]} but URL has /merge_requests/${mrIid}`,
    };
  }

  return { link, status: "ok", message: `MR !${mrIid} — URL format valid` };
}

export function verifyJira(link: LinkInfo): LinkResult {
  const m = link.url.match(/\/browse\/([A-Z]+-\d+)/);
  if (!m) {
    return {
      link,
      status: "error",
      message: "Could not parse Jira ticket URL",
    };
  }
  const ticket = m[1];

  const textTicket = link.link_text.match(/([A-Z]+-\d+)/);
  if (textTicket && textTicket[1] !== ticket) {
    return {
      link,
      status: "error",
      message: `Ticket mismatch: text says ${textTicket[1]} but URL has ${ticket}`,
    };
  }

  return { link, status: "ok", message: ticket };
}

// ---------------------------------------------------------------------------
// ReportLinkValidator
// ---------------------------------------------------------------------------

export class ReportLinkValidator {
  private agentRoot: string;
  private verbose: boolean;
  results: LinkResult[] = [];
  private githubToken: string | undefined;
  private gitlabToken: string | undefined;
  private engineerGithub: Map<string, string>;
  private engineerGitlab: Map<string, string>;

  constructor(agentRoot: string, verbose: boolean = false) {
    this.agentRoot = agentRoot;
    this.verbose = verbose;
    this.githubToken = process.env.GITHUB_PAT;
    this.gitlabToken = process.env.GITLAB_PAT;
    const configPath = resolve(this.agentRoot, "data", "team-config.json");
    const maps = loadEngineerMaps(configPath);
    this.engineerGithub = maps.github;
    this.engineerGitlab = maps.gitlab;
  }

  async verifyGithubPr(link: LinkInfo): Promise<LinkResult> {
    const m = link.url.match(/github\.com\/([^/]+)\/([^/]+)\/pull\/(\d+)/);
    if (!m) {
      return {
        link,
        status: "error",
        message: "Could not parse GitHub PR URL",
      };
    }

    const [, owner, repo, prNum] = m;

    const textNum = link.link_text.match(/#(\d+)/);
    if (textNum && textNum[1] !== prNum) {
      return {
        link,
        status: "error",
        message: `PR number mismatch: text says #${textNum[1]} but URL has /pull/${prNum}`,
      };
    }

    if (!this.githubToken) {
      return {
        link,
        status: "warning",
        message: "No GITHUB_PAT — skipped API check",
      };
    }

    const apiUrl = `https://api.github.com/repos/${owner}/${repo}/pulls/${prNum}`;
    const data = await apiGet(apiUrl, this.githubToken);

    if (data === null) {
      return {
        link,
        status: "error",
        message: `404 NOT FOUND: ${owner}/${repo}/pull/${prNum}`,
      };
    }

    const actualAuthor =
      ((data.user as Record<string, unknown>)?.login as string) ?? "unknown";
    if (link.engineer) {
      const expected = this.engineerGithub.get(link.engineer);
      if (expected && actualAuthor !== expected) {
        return {
          link,
          status: "error",
          message: `WRONG AUTHOR: expected ${expected}, got ${actualAuthor}`,
        };
      }
    }

    const actualTitle = (data.title as string) ?? "";
    let descInText = link.link_text.replace(/^(PR #\d+ - |MR \d+ - )/, "");
    descInText = descInText.replace(/\s*\([^)]*\)\s*$/, "").trim();
    if (
      descInText &&
      actualTitle &&
      !descInText.toLowerCase().includes(actualTitle.toLowerCase()) &&
      !actualTitle.toLowerCase().includes(descInText.toLowerCase())
    ) {
      return {
        link,
        status: "warning",
        message: `Title drift: text='${descInText}' vs actual='${actualTitle.slice(0, 60)}'`,
      };
    }

    const state = (data.state as string) ?? "unknown";
    const merged = (data.merged as boolean) ?? false;
    const statusStr = merged ? "merged" : state;
    return {
      link,
      status: "ok",
      message: `${owner}/${repo} #${prNum} (${actualAuthor}, ${statusStr})`,
    };
  }

  async verifyGitlabMr(link: LinkInfo): Promise<LinkResult> {
    const m = link.url.match(/gitlab[^/]*\/(.+?)\/-\/merge_requests\/(\d+)/);
    if (!m) {
      return {
        link,
        status: "error",
        message: "Could not parse GitLab MR URL",
      };
    }

    const [, projectPath, mrIid] = m;

    const textNum = link.link_text.match(/(?:MR |!)(\d+)/);
    if (textNum && textNum[1] !== mrIid) {
      return {
        link,
        status: "error",
        message: `MR number mismatch: text says !${textNum[1]} but URL has /merge_requests/${mrIid}`,
      };
    }

    if (!this.gitlabToken) {
      return {
        link,
        status: "warning",
        message: "No GITLAB_PAT — skipped API check",
      };
    }

    const encodedPath = projectPath.replace(/\//g, "%2F");
    const apiUrl = `https://gitlab.cee.redhat.com/api/v4/projects/${encodedPath}/merge_requests/${mrIid}`;
    const data = await apiGet(apiUrl, this.gitlabToken, "bearer");

    if (data === null) {
      return {
        link,
        status: "error",
        message: `404 NOT FOUND: ${projectPath}!${mrIid}`,
      };
    }

    const actualAuthor =
      ((data.author as Record<string, unknown>)?.username as string) ??
      "unknown";
    if (link.engineer) {
      const expected = this.engineerGitlab.get(link.engineer);
      if (expected && actualAuthor !== expected) {
        return {
          link,
          status: "error",
          message: `WRONG AUTHOR: expected ${expected}, got ${actualAuthor}`,
        };
      }
    }

    const state = (data.state as string) ?? "unknown";
    return {
      link,
      status: "ok",
      message: `${projectPath} !${mrIid} (${actualAuthor}, ${state})`,
    };
  }

  private printResult(result: LinkResult): void {
    if (result.status === "ok") {
      if (this.verbose) {
        console.log(
          `  ${GREEN}✓${RESET} Line ${result.link.line_num}: ${result.message}`,
        );
      }
    } else if (result.status === "warning") {
      console.log(
        `  ${YELLOW}⚠${RESET} Line ${result.link.line_num}: ${result.message}`,
      );
    } else {
      console.log(
        `  ${RED}✗${RESET} Line ${result.link.line_num}: ${result.link.link_text}`,
      );
      console.log(`    ${RED}${result.message}${RESET}`);
    }
  }

  private printSummary(): boolean {
    const total = this.results.length;
    const ok = this.results.filter((r) => r.status === "ok").length;
    const warnings = this.results.filter((r) => r.status === "warning").length;
    const errors = this.results.filter((r) => r.status === "error").length;

    console.log(`\n${BLUE}${"=".repeat(60)}${RESET}`);
    let summary = `Summary: ${total} checked, ${GREEN}${ok} passed${RESET}, `;
    if (warnings) {
      summary += `${YELLOW}${warnings} warnings${RESET}, `;
    }
    if (errors) {
      summary += `${RED}${errors} broken${RESET}`;
    } else {
      summary += `${GREEN}0 broken${RESET}`;
    }
    console.log(summary);
    console.log(`${BLUE}${"=".repeat(60)}${RESET}\n`);

    return errors === 0;
  }

  async validate(filePath: string): Promise<boolean> {
    console.log(`\n${BLUE}${"=".repeat(60)}${RESET}`);
    console.log(`${BLUE}Report Link Validation${RESET}`);
    console.log(`${BLUE}${"=".repeat(60)}${RESET}`);
    console.log(`\nValidating: ${filePath}\n`);

    const content = readFileSync(filePath, "utf-8");
    const links = parseLinks(content);
    const classified = classifyLinks(links);

    if (classified.github_pr.length > 0) {
      console.log(
        `\n${BLUE}GitHub PRs (${classified.github_pr.length})${RESET}`,
      );
      for (const link of classified.github_pr) {
        const result = await this.verifyGithubPr(link);
        this.results.push(result);
        this.printResult(result);
      }
    }

    if (classified.gitlab_mr.length > 0) {
      console.log(
        `\n${BLUE}GitLab MRs (${classified.gitlab_mr.length})${RESET}`,
      );
      for (const link of classified.gitlab_mr) {
        const result = await this.verifyGitlabMr(link);
        this.results.push(result);
        this.printResult(result);
      }
    }

    if (classified.jira.length > 0) {
      console.log(`\n${BLUE}Jira Tickets (${classified.jira.length})${RESET}`);
      for (const link of classified.jira) {
        const result = verifyJira(link);
        this.results.push(result);
        this.printResult(result);
      }
    }

    return this.printSummary();
  }
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function findLatestReport(agentRoot: string): string | null {
  const reportDir = resolve(agentRoot, "data", "output");
  if (!existsSync(reportDir)) return null;
  const reports = readdirSync(reportDir)
    .filter((f) => f.startsWith("weekly-update-") && f.endsWith(".md"))
    .sort()
    .reverse();
  return reports.length > 0 ? resolve(reportDir, reports[0]) : null;
}

// ---------------------------------------------------------------------------
// CLI
// ---------------------------------------------------------------------------

async function main(): Promise<void> {
  const args = process.argv.slice(2);
  let filePath: string | null = null;
  let verbose = false;

  for (let i = 0; i < args.length; i++) {
    if (args[i] === "--verbose" || args[i] === "-v") {
      verbose = true;
    } else if (!args[i].startsWith("-")) {
      filePath = args[i];
    }
  }

  const scriptDir = dirname(resolve(process.argv[1]));
  const agentRoot = resolve(scriptDir, "..");

  if (filePath) {
    if (!filePath.startsWith("/")) {
      filePath = resolve(process.cwd(), filePath);
    }
  } else {
    filePath = findLatestReport(agentRoot);
    if (!filePath) {
      console.log(
        `${RED}No weekly reports found in agents/weekly-team-update/data/output/${RESET}`,
      );
      process.exit(1);
    }
  }

  if (!existsSync(filePath)) {
    console.log(`${RED}File not found: ${filePath}${RESET}`);
    process.exit(1);
  }

  const validator = new ReportLinkValidator(agentRoot, verbose);
  const success = await validator.validate(filePath);
  process.exit(success ? 0 : 1);
}

const isMain =
  process.argv[1] &&
  resolve(process.argv[1]) === resolve(import.meta.url.replace("file://", ""));
if (isMain) {
  void main();
}
